core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,196 @@
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+    loader: L,
+}
+
+impl<L: VmRunnerStorageLoader> VmRunnerStorage<L> {
+    /// Creates a new VM runner storage using provided Postgres pool and RocksDB path.
+    pub async fn new(
+        pool: ConnectionPool<Core>,
+        rocksdb_path: String,
+        loader: L,
+        enum_index_migration_chunk_size: usize, // TODO: Remove
+        validation_computational_gas_limit: u32,
+        chain_id: L2ChainId,
+    ) -> anyhow::Result<(Self, AsyncCatchupTask)> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let task = AsyncCatchupTask::new(
+            pool.clone(),
+            rocksdb_path,
+            enum_index_migration_chunk_size,
+            rocksdb_cell.clone(),
+            Some(loader.latest_processed_batch(&mut conn).await?),
+        );
+        drop(conn);
+        Ok((
+            Self {
+                pool,
+                l1_batch_params_provider,
+                validation_computational_gas_limit,
+                chain_id,
+                rocksdb_cell,
+                loader,
+            },
+            task,
+        ))
+    }
+
+    async fn access_storage_inner(
+        &self,
+        stop_receiver: &watch::Receiver<bool>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
+        if let Some(rocksdb) = self.rocksdb_cell.get() {
+            let mut connection = self
+                .pool
+                .connection_tagged(L::name())
+                .await
+                .context("Failed getting a Postgres connection")?;
+            PgOrRocksdbStorage::access_storage_rocksdb(
+                &mut connection,
+                rocksdb.clone(),
+                stop_receiver,
+                l1_batch_number,
+            )
+            .await
+            .context("Failed accessing RocksDB storage")
+        } else {
+            Ok(Some(
+                PgOrRocksdbStorage::access_storage_pg(&self.pool, l1_batch_number)
+                    .await
+                    .context("Failed accessing Postgres storage")?,
+            ))
+        }
+    }
+
+    /// Loads next unprocessed L1 batch along with all transactions that VM runner needs to
+    /// re-execute. These are the transactions that are included in a sealed miniblock belonging
+    /// to a sealed L1 batch (with state keeper being the source of truth). The order of the
+    /// transactions is the same as it was when state keeper executed them.
+    ///
+    /// Can return `None` if there are no batches to be processed.
+    ///
+    /// # Errors
+    ///
+    /// Propagates DB errors.
+    pub async fn load_next_batch(&self) -> anyhow::Result<Option<BatchData>> {
```

Comment:
I was thinking that this component wouldn't only serve as a provider but would continuously load data for new batches (loader should provide info about max batch ready to be loaded) and drop data that is not needed. I think it will be better than having this provider struct and actual updater separately.